sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -608,6 +608,12 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
   if (NumDevices)
     *NumDevices = ZeDeviceCount;
 
+  if (Devices == nullptr) {
+    assert(NumEntries == 0 &&
+           "NumEntries should be zero when querying the number of devices");
+    return PI_SUCCESS;
+  }
+
   try {
     // TODO: Delete array at teardown
     ze_device_handle_t *ZeDevices = new ze_device_handle_t[ZeDeviceCount];
```

Comment:
Is `ZeDevices` deleted anywhere?
Why comment says it should be deleted as teardown? Unless I missing something, I think only memory allocated for `Devices` below must be deallocated by the caller of `piDevicesGet`, but this array can be deleted right after the for loop.

```suggestion
    vector<ze_device_handle_t> ZeDevices(ZeDeviceCount);
```

Reply:
This code indeed has memory leak (I missed this part :))
This memory is used create _pi_device object.
The _pi_device constructor contains ZeDevice member, which contains this memory.
So, we cannot deallocate this memory right after the loop.
I will work this memory leak issue separately (separate PR) if it is okay.

Reply:
I'm okay to resolve this memory leak separately, but I think you can just apply suggested change to fix it.

> The _pi_device constructor contains ZeDevice member, which contains this memory.
> So, we cannot de-allocate this memory right after the loop.

Why? The constructor is called inside the loop and I don't see any uses of `ZeDevices` outside of it. Am I missing something?

It won't resolve another memory leak for allocations on line 624.

Reply:
The result of constructor _pi_device is assigned to the variable "Devices".
"Devices" is passed as reference to this function.
So, the caller of this function will use "Devices" objects created in this function.
If we deallocate ZeDevice member, the caller uses corrupted "Devices" objects.

Reply:
> The result of constructor _pi_device is assigned to the variable "Devices".
> "Devices" is passed as reference to this function.
> So, the caller of this function will use "Devices" objects created in this function.
> If we deallocate ZeDevice member, the caller uses corrupted "Devices" objects.

No. `_pi_device` constructor accepts handler object by value, so each `Devices[I]` will have each own copy of handler object and original array can be safely deleted.

Reply:
@smaslov-intel , what do you think?

Reply:
I just changed the code to call "delete" instead of making vector as you suggested.
The reason is that zeDebiceGet does not take vector as parameter.

Reply:
You can use `std::vector::data()` method to get a pointer to the data to pass it to `zeDeviceGet`.
Using `std::vector` is better than explicitly calling `delete` for two reasons:
1. If `new _pi_device` throws an exception, `delete` won't be called - memory leak
2. If `Devices[I]-initialize()` return code is not `PI_SUCCESS`, `delete` won't be called - memory leak.

Reply:
ok. I changed as you requested.

Reply:
Sorry that I lost track of this discussion, the resolution looks good to me!